### ISSUE TITLE
HTTP proxy not honored for all requests

### ISSUE THREAD

**Issue by DaAwesomeP** [PaperMC/Waterfall#478]
I am unsure if this is an issue with vanilla or Waterfall or it is easily patched. The server is running behind a firewall that requires all requests go through an HTTP proxy. I added the following Java flags:
```
-Dhttp.proxyHost=my.proxy.com -Dhttp.proxyPort=3128 -Dhttps.proxyHost=my.proxy.com -Dhttps.proxyPort=3128 -Dhttp.nonProxyHosts="localhost|127.*|[::1]"
```
Some HTTP requests are honored through the firewall and others are not (many from plugins doing version checks). I assume this is from different ways of sending HTTP/writing TCP. Most importantly, Mojang user authentication fails:
```
io.netty.channel.AbstractChannel$AnnotatedConnectException: finishConnect(..) failed: Connection refused: sessionserver.mojang.com/99.84.124.180:443
```
I can verify that the proxy indeed works:
```
$ curl https://sessionserver.mojang.com --proxy my.proxy.com:3128
{"Status":"OK","Runtime-Mode":"productionMode","Application-Author":"Mojang Web Force","Application-Description":"Mojang Session Server.","Specification-Version":"3.6.0","Application-Name":"yggdrasil.session.restlet.server","Implementation-Version":"3.6.0_build103","Application-Owner":"Mojang"}
```

**Comment by aikar** [comment]
Have you tried with a SOCKS5 Proxy? socksProxyHost? Should be a bit more reliable. Not sure other wise.

**Comment by DaAwesomeP** [comment]
This is an enterprise environment where unfortunately only the HTTP proxy is available. I could adapt the HTTP proxy to a SOCKS proxy locally, but that is not very direct, so I would like to avoid that if possible.

**Comment by electronicboy** [comment]
This is down to netty, bungee, and by extension waterfall, don't appear to
actually handle proxies in the connection to mojangs auth servers

**Comment by DaAwesomeP** [comment]
OK, I just tried using `socksify` (from Dante package) to no avail. It basically routes all traffic of the program run in it through a proxy. Worked for all the programs I could try it with except Java.

**Comment by DaAwesomeP** [comment]
OK, we found a workaround, and its a bit ugly, but it works. We made a custom patch and changed `sessionserver.mojang.com` to a local server. That server is a reverse proxy that uses the upstream outgoing proxy. It seems to work fine and does not create any security risks.

Proxychains, socksify, and a transparent proxy with iptables all failed for various reasons. Notably that this is a CloudFront CDN address with a shifting IP, and Java would always resolve the IP _before accessing/forwarding_. So the outgoing proxy would receive an IP, not a host, and that IP is not whitelisted.

We found the culprit of the resolving and the lack of using the proxy here: https://github.com/SpigotMC/BungeeCord/blob/master/proxy/src/main/java/net/md_5/bungee/http/HttpClient.java. We have something functioning at the moment so we need to move to other issues at the moment, but a better long-term solution would be to patch this file to use the proxy and resolve as expected. If we get there, I will share the results (and patch).
